### PR TITLE
Feature/sonos auth final

### PR DESCRIPTION
**Pull request by Nacrylic**
Fix URL security problem with Sonos, on covertArt Sonos change size parameters, skip size parameter for validate security.

### CODE REVIEW

**Comment by ice954pw** [comment]
So far the single user plugin works great, however, you can't search when you add anonymous connection the library on sonos app, which makes the need for the multi user so it has an account. However, registration works under multi user, except it asks you to login to airsonic, then it complains and throws back SQL Errors in the logs. This was the only thing I was irritated that didn't work and you've done a great job getting it to. Any insight to the SQL Errors, it's saying the tables aren't created but it looks to me by generating the pull they should've been?

**Comment by ice954pw** [comment]
`t org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179) ~[cxf-core-3.2.6.jar!/:3.2.6]
        at org.apache.cxf.jaxws.JAXWSMethodInvoker.performInvocation(JAXWSMethodInvoker.java:66) ~[cxf-rt-frontend-jaxws-3.2.6.jar!/:3.2.6]
        at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96) ~[cxf-core-3.2.6.jar!/:3.2.6]
        ... 112 common frames omitted
Caused by: java.sql.SQLException: Table not found in statement [select username, householdid, linkcode from sonoslink where linkcode=?]
        at org.hsqldb.jdbc.Util.throwError(Unknown Source) ~[hsqldb-1.8.0.7.jar!/:private-2006/09/24-10:30:15]
        at org.hsqldb.jdbc.jdbcPreparedStatement.<init>(Unknown Source) ~[hsqldb-1.8.0.7.jar!/:private-2006/09/24-10:30:15]
        at org.hsqldb.jdbc.jdbcConnection.prepareStatement(Unknown Source) ~[hsqldb-1.8.0.7.jar!/:private-2006/09/24-10:30:15]
        at org.springframework.jdbc.core.JdbcTemplate$SimplePreparedStatementCreator.createPreparedStatement(JdbcTemplate.java:1526) ~[spring-jdbc-4.3.21.RELEASE.jar!/:4.3.21.RELEASE]
        at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:633) ~[spring-jdbc-4.3.21.RELEASE.jar!/:4.3.21.RELEASE]
        ... 139 common frames omitted
`

**Comment by Nacrylic** [comment]
Sorry, I forgot to specify that merge must be apply on the "Fix the registration of Sonos service and link with security" :  https://github.com/airsonic/airsonic/pull/863

In that you are the creation of the table. It is possible to link this pull request on other pull request?

**Comment by ice954pw** [comment]
Let me know I'll merge I actually have it working well. It's using an account and I'm streaming to it now :)



*sent from mobile*
Patrick Winkel
Information Security
Digital Forensics
Enterprise Computing
CDS LLC.
P.O. Box 9568, New Haven, CT 06535
P: 203.776.2637<tel:203.776.2637> ext 16
C: 203.980.8807<tel:203.980.8807>
F: 203.776.1198<tel:203.776.1198>
www.cdsnewhaven.com<http://www.cdsnewhaven.com>


-------- Original message --------

**Comment by muff1nman** [comment]
So does this supercede #852 and #863? Can those be closed and is this ready for review?

**Comment by Nacrylic** [comment]
They are all to be reviewed and must be kept. Do you need something? Help?

**Comment by zombiehoffa** [comment]
it would be super awesome if you guys could resolve the conflicts and get this working again...

**Comment by jvoisin** [comment]
Unfortunately, I don't think that any core developer is using Sonos device. Someone else would have to do it.

**Comment by Nacrylic** [comment]
Okay, I'll fix all the conflicts and do what I said in another post : 

> So I re-evaluate that code, this service must not be get by component scan but declared in global security. And it will be move to the security package.
> 
> All initialisation about security it's set there and the service just use the global component.
> 

Also, Sonos provides a set of PHP files to test the services. It was my intention, once the sources were merged, to provide the set with the configuration files made so that a developer could perform tests without a Sonos system.

But before, there was another bug I wanted to fix: https://github.com/airsonic/airsonic/issues/889, but I was waiting for the code to be fixed.

Let me know if it suits you!

**Comment by ice954pw** [comment]
Look forward to this being fixed it's functional right now on my side build but could definitely be better. Sonos stopped searching it randomly but I am still able to manually browse. Just a bug figured I'd mention but I bet it's because of the code not announcing.



*sent from mobile*
Patrick Winkel
Information Security
Digital Forensics
Enterprise Computing
CDS LLC.
P.O. Box 9568, New Haven, CT 06535
P: 203.776.2637<tel:203.776.2637> ext 16
C: 203.980.8807<tel:203.980.8807>
F: 203.776.1198<tel:203.776.1198>
www.cdsnewhaven.com<http://www.cdsnewhaven.com>


-------- Original message --------

**Comment by ice954pw** [comment]
Side note been poking around writing an Alexa skill for voice control I think it would be a really neat feature.

**Comment by zombiehoffa** [comment]
You guys are awesome.

**Comment by Nacrylic** [comment]
I know that, but I don't want the code to remain in pull request after redoing everything and in 2 or 3 months I have to start again.

**Comment by jvoisin** [comment]
@Nacrylic I'll be more than happy to review, test and merge your PR, if there is a (easy) way for me to test this without having a Sonos device :)

**Comment by ice954pw** [comment]
I'll gladly test this and merge the pull with the new version if you want, I have several sonos speakers and amps.

**Comment by ice954pw** [comment]
Let me know when it's ready

File: airsonic-main/src/test/java/org/airsonic/player/service/sonos/sonos_manuel_test_.http [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#======================================================================================================
+# For manuel test
+#
```
Comment (eharris) [inline airsonic-main/src/test/java/org/airsonic/player/service/sonos/sonos_manuel_test_.http]: The proper spelling is "manual".  The filename should be corrected to "sonos_manual_test.h" as well.

File: airsonic-main/src/main/java/org/airsonic/player/service/SecurityService.java [hunk 2]
```diff
@@ -336,15 +372,101 @@ protected boolean isFileInFolder(String file, String folder) {
...
+    private String createLinkCode() {
+        return UUID.randomUUID().toString().replace("-", "");
     }
 
-    public void setUserCache(Ehcache userCache) {
-        this.userCache = userCache;
+
+    //--------------------------------------------------------------------------------------------------------
+
+
```
Comment (eharris) [inline airsonic-main/src/main/java/org/airsonic/player/service/SecurityService.java]: nit: so much whitespace is probably unnecessary

File: airsonic-main/src/main/java/org/airsonic/player/security/GlobalSecurityConfig.java [hunk 3]
```diff
@@ -67,6 +67,7 @@ public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception
             settingsService.save();
         }
         auth.authenticationProvider(new JWTAuthenticationProvider(jwtKey));
+
```
Comment (eharris) [inline airsonic-main/src/main/java/org/airsonic/player/security/GlobalSecurityConfig.java]: nit: additional whitespace probably isn't needed

File: airsonic-main/src/main/java/org/airsonic/player/service/SettingsService.java [hunk 4]
```diff
@@ -212,6 +221,7 @@
     private static final Integer DEFAULT_DATABASE_MYSQL_VARCHAR_MAXLENGTH = 512;
     private static final String DEFAULT_DATABASE_USERTABLE_QUOTE = null;
 
+
     // Array of obsolete keys.  Used to clean property file.
```
Comment (eharris) [inline airsonic-main/src/main/java/org/airsonic/player/service/SettingsService.java]: nit: Unnecessary whitespace

File: airsonic-main/src/main/java/org/airsonic/player/service/SettingsService.java [hunk 5]
```diff
@@ -1444,4 +1459,45 @@ public void resetDatabaseToDefault() {
     String getPlaylistExportFormat() {
         return getProperty(KEY_EXPORT_PLAYLIST_FORMAT, DEFAULT_EXPORT_PLAYLIST_FORMAT);
     }
+
+
+    public String getHost() {
```
Comment (eharris) [inline airsonic-main/src/main/java/org/airsonic/player/service/SettingsService.java]: nit: Too much whitespace

File: airsonic-main/src/main/java/org/airsonic/player/service/sonos/SonosLinkkInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
+package org.airsonic.player.service.sonos;
+
+
+import com.google.common.collect.Sets;
```
Comment (eharris) [inline airsonic-main/src/main/java/org/airsonic/player/service/sonos/SonosLinkkInterceptor.java]: nit: unnecessary whitespace

File: airsonic-main/src/main/java/org/airsonic/player/service/sonos/SonosServiceRegistration.java [hunk 7]
```diff
@@ -35,19 +36,55 @@
...
+            this.fieldValue = fieldValue;
+        }
+
+        public String getFieldValue() {
+            return fieldValue;
+        }
+    }
+
+
     /**
```
Comment (eharris) [inline airsonic-main/src/main/java/org/airsonic/player/service/sonos/SonosServiceRegistration.java]: nit: unnecessary extra whitespace

File: airsonic-main/src/main/resources/org/airsonic/player/i18n/ResourceBundle_fr.properties [hunk 8]
```diff
@@ -28,7 +28,6 @@ login.logout=Vous \u00EAtes maintenant d\u00E9connect\u00E9.
 login.error=Erreur de nom d''utilisateur ou de mot de passe.
 login.insecure={0} n''est pas s\u00E9curis\u00E9. Connectez-vous avec &quot;admin&quot; comme nom d''utilisateur<br> et mot de passe, puis modifiez le imm\u00E9diatement.
 login.recover=Mot de passe oubli\u00E9 ?
-
```
Comment (eharris) [inline airsonic-main/src/main/resources/org/airsonic/player/i18n/ResourceBundle_fr.properties]: nit: this whitespace should be preserved

File: airsonic-main/src/main/resources/org/airsonic/player/theme/2010.properties [hunk 9]
```diff
@@ -12,33 +12,33 @@ detailColor = AAAAAA
...
-ratingHalfImage = icons/2010/rating_half.png
-removeImage = icons/2010/remove.png
-searchImage = icons/2010/search.png
-settingsImage = icons/2010/settings.png
-statusImage = icons/2010/status.png
-upImage = icons/2010/up.png
-uploadImage = icons/2010/upload.png
-volumeImage = icons/2010/now_playing.png
-wapImage = icons/2010/phone.png
+addImage = /icons/2010/add.png
```
Comment (eharris) [inline airsonic-main/src/main/resources/org/airsonic/player/theme/2010.properties]: Why are all these icon references being changed to absolute?

File: airsonic-main/src/main/java/org/airsonic/player/service/SecurityService.java [hunk 10]
```diff
@@ -448,15 +450,23 @@ public void setSonosUser(String sonosLinkToken) throws SonosSoapFault.LoginUnaut
...
+        setUser(SONOS_USER_USERNAME);
+    }
+
+    private void setUser(String username) throws SonosSoapFault.LoginUnauthorized {
+        User user = getUserByName(SONOS_USER_USERNAME, true);
+        Authentication authentication = authenticate(user.getUsername(), user.getPassword());
+        SecurityContextHolder.getContext().setAuthentication(authentication);
+    }
+
+        // must be strict 32 character limit on length
```
Comment (eharris) [inline airsonic-main/src/main/java/org/airsonic/player/service/SecurityService.java]: nit: fix indent

File: airsonic-main/src/main/java/org/airsonic/player/service/SettingsService.java [hunk 11]
```diff
@@ -1461,6 +1466,14 @@ String getPlaylistExportFormat() {
     }
 
 
+    public String getSonosLinkMethod(){
```
Comment (eharris) [inline airsonic-main/src/main/java/org/airsonic/player/service/SettingsService.java]: nit: unnecessary extra whitespace

File: airsonic-main/src/main/webapp/WEB-INF/jsp/sonosSettings.jsp [hunk 12]
```diff
@@ -72,12 +81,23 @@
...
+    <p></p>
+    <h2 <c:if test="${empty model.returnCodes}">hidden</c:if>><fmt:message key="sonossettings.returnCodes"/></h2>
+    <p></p>
+    <div style="padding:5" class="log indent" <c:if test="${empty model.returnCodes}">hidden</c:if>>
+        <c:forEach items="${model.returnCodes}" var="returnCode">
+            <p><fmt:message key="${returnCode}"/></p>
+        </c:forEach>
+    </div>
+
+
```
Comment (eharris) [inline airsonic-main/src/main/webapp/WEB-INF/jsp/sonosSettings.jsp]: nit: unnecessary extra whitespace

**Comment by eharris** [comment]
@Nacrylic I would strongly suggest that you rebase your changes to get rid of the changes that you have backed out, which would make it much easier to review.  It seems that big parts of the changes have been modified (and in some cases reverted) multiple times in these commits without good reason.

**Comment by eharris** [comment]
@Nacrylic I see you closed this PR. Are you planning to open a new one with the cleaned up submission?

**Comment by jvoisin** [comment]
This new PR is likely #1177
